zebrad/src/components/sync.rs

```diff
@@ -583,65 +524,17 @@ where
     async fn request_blocks(&mut self, hashes: Vec<block::Hash>) -> Result<(), Report> {
         tracing::debug!(hashes.len = hashes.len(), "requesting blocks");
         for hash in hashes.into_iter() {
-            // Avoid re-downloading blocks that have already been verified.
-            // This is particularly important for nodes on slow or unreliable
-            // networks.
+            // If we've queued the download of a hash behind our current chain tip,
+            // we've been given bad responses by our peers.  Abort the sync and restart.
+            tracing::debug!(?hash, "checking if state contains hash");
             if self.state_contains(hash).await? {
```

Comment:
I'd like to reduce the number of state calls. Here are two possible optimisations:

1. Check the state's tip height once per request_blocks. Reject any blocks that have a coinbase height that is lower than the tip height minus the reorg limit. They can't possibly be valid, because we already have a finalised block for that height.

2. Check if the download set has queued their hash - if it has, ignore the request. Then check the state, and restart if the hash is in the state. If the block is in the download set, it can't possibly be in the state, because a previous state check would have caught it.

Reply:
I'm not sure either of those optimizations work — in case (1), we don't have access to heights, only hashes, so I'm not sure we can do that query, and in case (2), we almost never duplicate download requests, so we wouldn't save much.

I think a better solution would be along the lines of the one I mentioned in my comment on this rebase, to eliminate state checks from extendtips invocations.

Reply:
Ok, sounds good.

Have you made those changes to ExtendTips yet?
Or are they in another ticket?

Reply:
Ah, I just saw 5477ce141580f108ff711bdf0913886b077b35db